core/app/[locale]/(default)/product/[slug]/_actions/wishlist-action.ts

```diff
@@ -0,0 +1,304 @@
+  } catch {
+    ctx.addIssue({ code: 'custom', message: 'Invalid menuItem payload.' });
+
+    return z.NEVER;
+  }
+});
+
+const schema = z.object({
+  productId: z.number().nonnegative().min(1),
+  selectedSku: z.string(),
+  menuItem: menuItemParser,
+});
+
+interface WishlistAddMutationVariables {
+  wishlistId: number;
+  productId: number;
+  variantId?: number;
+}
+
+interface WishlistRemoveMutationVariables {
+  wishlistId: number;
+  wishlistItemId: number;
+}
+
+const getVariantIdFromSku = cache(
+  async (productId: number, sku: string, customerAccessToken?: string) => {
+    const { data } = await client.fetch({
+      document: VariantIdFromSkuQuery,
+      variables: { productId, sku },
+      customerAccessToken,
+      fetchOptions: { next: { revalidate } },
+    });
+
+    return data.site.product?.variants.edges?.[0]?.node.entityId ?? undefined;
+  },
+);
+
+async function addToDefaultWishlist(
+  customerAccessToken: string,
+  wishlistName: string,
+  productId: number,
+  variantId?: number,
+) {
+  const { data } = await client.fetch({
+    document: CreateWishlistMutation,
+    variables: {
+      input: {
+        name: wishlistName,
+        isPublic: false,
+        items: [{ productEntityId: productId, variantEntityId: variantId }],
+      },
+    },
+    customerAccessToken,
+    fetchOptions: { cache: 'no-store' },
+  });
+
+  if (!data.wishlist.createWishlist?.result) {
+    throw new Error('Failed to add item to default wishlist. Response was empty.');
+  }
+}
+
+async function addToWishlist(customerAccessToken: string, variables: WishlistAddMutationVariables) {
+  const { data } = await client.fetch({
+    document: AddToWishlistMutation,
+    variables,
+    customerAccessToken,
+    fetchOptions: { cache: 'no-store' },
+  });
+
+  if (!data.wishlist.addWishlistItems?.result) {
+    throw new Error('Failed to add item to wishlist. Response was empty.');
+  }
+}
+
+async function removeFromWishlist(
+  customerAccessToken: string,
+  variables: WishlistRemoveMutationVariables,
+) {
+  const { data } = await client.fetch({
+    document: DeleteWishlistItemsMutation,
+    variables,
+    customerAccessToken,
+    fetchOptions: { cache: 'no-store' },
+  });
+
+  if (!data.wishlist.deleteWishlistItems?.result) {
+    throw new Error('Failed to add item to wishlist. Response was empty.');
+  }
+}
+
+function getLoginRedirect(redirectTo: string) {
+  const loginParams = new URLSearchParams({ redirectTo });
+
+  return `/login?${loginParams.toString()}`;
+}
+
+export async function wishlistAction(payload: FormData): Promise<void> {
+  const locale = await getLocale();
+  const customerAccessToken = await getSessionCustomerAccessToken();
+  const t = await getTranslations('Product.ProductDetails.WishlistButton');
+  const submission = parseWithZod(payload, { schema });
+
+  if (submission.status !== 'success') {
+    await serverToast.error(t('Errors.unexpected'));
+
+    return;
+  }
```

Comment:
We usually just response with a `reply()` in these cases because it means the form was invalid. In this case, the form should never be invalid since we control the form, but still, we should probably propagate as a reply.

```
  if (submission.status !== 'success') {
    return { lastResult: submission.reply() };
  }
```

Reply:
@jorgemoya We don't use any sort of form states here. This is a void promise, so we don't need to return any sort of submission or result - all we need to do is fire off the toast messages

Reply:
That makes sense, thanks!